pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go

```diff
@@ -282,20 +282,33 @@ func syncClaim(volumeIndex *persistentVolumeOrderedIndex, binderClient binderCli
 			return fmt.Errorf("A volume match does not exist for persistent claim: %s", claim.Name)
 		}
 
+		// create a reference to the claim and assign it to the volume being bound.
+		// the volume is a pointer and assigning the reference fixes a race condition where another
+		// claim might match this volume but before the claimRef is persistent in the next case statement
+		claimRef, err := api.GetReference(claim)
+		if err != nil {
+			return fmt.Errorf("Unexpected error getting claim reference: %v\n", err)
+		}
+
 		// make a binding reference to the claim.
-		// triggers update of the claim in this controller, which builds claim status
+		volume.Spec.ClaimRef = claimRef
```

Comment:
I see what you're getting at and I believe it will work.

Your questions answered:

1.  No update/double call to sync.  volumeIndex.Update just puts the object in the store's map and indexers.  It's not a watch event.

2.  There is a common mutex between all updates, so another claim would only be waiting to find a volume.  By manually updating the volume in the index with the version returned from the API server, that volume's current state is already "bound" and won't be paired with any other claim.

I'll get change done and ping when ready.  Thanks for the review :)